File: app/src/main/java/com/example/daniel/currencyconverter/Currency/CurrencyModel.java
package com.example.daniel.currencyconverter.Currency;

import android.util.Log;

import com.example.daniel.currencyconverter.File.FileRepository;
import com.example.daniel.currencyconverter.Network.NetworkRepository;
import com.example.daniel.currencyconverter.Util.NetworkUtil;
import com.example.daniel.currencyconverter.XmlResponseModels.Envelope;

import io.reactivex.Observable;

/**
 * Created by Daniel on 2017-11-14.
 */

public class CurrencyModel implements MainContract.Model {
    private NetworkRepository networkRepository;
    private FileRepository fileRepository;
    private NetworkUtil networkUtil;

    public CurrencyModel(NetworkRepository networkRepository,
                         FileRepository fileRepository,
                         NetworkUtil networkUtil){
        this.networkRepository = networkRepository;
        this.fileRepository = fileRepository;
        this.networkUtil = networkUtil;
    }


    @Override
    public Observable<Envelope> getCurrency() {
        Envelope envelope = fileRepository.readFromFile();
        if(envelope == null){
            return networkRepository.getResponse().map(envelope1 -> fileRepository.saveToFile(envelope1));
        }
        if(!networkUtil.haveConnection()){
            Log.d("ModelCurrency", "No connection, reading data from file");
            return Observable.just(fileRepository.readFromFile());
        } else {
            long downloadedTime = envelope.getTimestamp();
            long timeNow = System.currentTimeMillis();
            if(networkUtil.haveWifiConnection()){
                Log.d("ModelCurrency", "wifi connection");

                if(( (timeNow/1000)/60) - ((downloadedTime/1000)/60) > 10){
                    Log.d("ModelCurrency", "Downloading using wifi");

                    return networkRepository.getResponse().map(envelope1 -> fileRepository.saveToFile(envelope1));
                }
            } else if(networkUtil.haveWifiMobile()){
                Log.d("ModelCurrency", "mobile connection");

                if(( (timeNow/1000)/60) - ((downloadedTime/1000)/60) > 60){
                    Log.d("ModelCurrency", "Downloading using mobile data");
                    return networkRepository.getResponse().map(envelope1 -> fileRepository.saveToFile(envelope1));
                }
            }


        }
        Log.d("ModelCurrency", "reading from file");
        return Observable.just(fileRepository.readFromFile());
    }
}

File: app/src/main/java/com/example/daniel/currencyconverter/Currency/CurrencyModel.kt
package com.example.daniel.currencyconverter.Currency

import android.util.Log
import com.example.daniel.currencyconverter.File.FileRepository
import com.example.daniel.currencyconverter.Network.NetworkRepository
import com.example.daniel.currencyconverter.Util.NetworkUtil
import com.example.daniel.currencyconverter.XmlResponseModels.Envelope
import io.reactivex.Single

/**
 * Created by Daniel on 2017-11-14.
 */

class CurrencyModel(private val networkRepository: NetworkRepository,
                    private val fileRepository: FileRepository,
                    private val networkUtil: NetworkUtil) : MainContract.Model {


    override fun getCurrency(): Single<Envelope>? {
        val envelope = fileRepository.readFromFile()
        if (envelope == null) {
            if (networkUtil.haveConnection()) {
                val envelopeSingle = networkRepository.response
                        .map { envelope1 -> fileRepository.saveToFile(envelope1) }
                Log.d("ModelCurrency", " " + envelopeSingle.toString())

            }
        }
        if (!networkUtil.haveConnection()) {
            Log.d("ModelCurrency", "No connection")
            return null
        } else {
            val downloadedTime = envelope!!.timestamp
            val timeNow = System.currentTimeMillis()
            if (networkUtil.haveWifiConnection()) {
                Log.d("ModelCurrency", "wifi connection")

                if (timeNow / (1000 * 60) - downloadedTime / (1000 * 60) > 10) {
                    networkRepository.response
                }
            } else if (networkUtil.haveWifiMobile()) {
                Log.d("ModelCurrency", "mobile connection")

                if (timeNow / (1000 * 60) - downloadedTime / (1000 * 60) > 60) {
                    networkRepository.response
                }
            }


        }
        return networkRepository.response
    }
}
